fix epand_events for future events

frisi opened the pull request:
this is a follow up of  #262 for the 3.2.x branch (currently used for plone 5.2) that also adds a small test with explanation what the change is good for.

this fixes https://github.com/plone/plone.app.event/issues/261

mister-roboto commented:
@frisi thanks for creating this Pull Request and help improve Plone!

To ensure that these changes do not break other parts of Plone, the Plone test suite matrix needs to pass.

Whenever you feel that the pull request is ready to be tested, either start all jenkins jobs pull requests by yourself, or simply add a comment in this pull request stating:

```
@jenkins-plone-org please run jobs
```

With this simple comment all the jobs will be started automatically.

Happy hacking!

frisi commented:
@jenkins-plone-org please run jobs

frisi commented:
@thet @jensens could you please review this? thanks!

jensens commented on plone/app/event/base.py:
```diff
@@ -279,8 +279,12 @@ def expand_events(events, ret_mode,
-            occurrences = [_obj_or_acc(occ, ret_mode) for occ in
-                           IRecurrenceSupport(obj).occurrences(start, end)]
+
+            occ_set = set(IRecurrenceSupport(obj).occurrences(start, end))
```
If a set is used  then order is gone. If this is not an issue here, this is fine.

jensens replied:
@frisi I would like to get this in, but what do you think here?

frisi replied:
sorry for the delay, i was on holidays ;-)

good point @jensens 

personally i supply the sort parameter to  `expand_events(..sort='start')` all the time. so after obtaining all occurrences, they get sorted [here](https://github.com/plone/plone.app.event/blob/4.0.0a4/plone/app/event/base.py#L290-L291)

also plone.app.event passes `sort='start'` in [portlet_calendar.py](https://github.com/plone/plone.app.event/blob/4.0.0a4/plone/app/event/portlets/portlet_calendar.py#L226-L230), portlet_events, and [event_listing.py](https://github.com/plone/plone.app.event/blob/4.0.0a4/plone/app/event/browser/event_listing.py#L148)

dunno if other people rely on the occurrences being sorted correctly, though.
at least neither the docs of [`p.a.event.recurrences.RecurrenceSupport`](https://github.com/plone/plone.app.event/blob/4.0.0a4/plone/app/event/recurrence.py#L39) nor [`plone.event.recurrence.recurrence_sequence_ical`](https://github.com/plone/plone.event/blob/1.4.1/plone/event/recurrence.py#L18) explicitly state anything about the order of results.


not sure why i used a set here at all. maybe to make sure the object is not added twice w/o the need to iterate over the list.
but this should be covered by the condition. so using a list (and therefore keep the original order) and append the object to the lists should be fine as well.
will add a commit soon.. ;-)

jensens reviewed (approved):
See my comment, I do not know off head if this is an issue.
Otherwise LGTM.

frisi commented:
thanks for the heads-up @jensens 
i addressed your comment. 

if tests pass commits can be squashed and merged

@jenkins-plone-org please run jobs

frisi commented:
@jensens this should be good to merge now, right?
